### DISCUSSION TITLE
Confusion around behaviour when importing server components into client components

### DISCUSSION THREAD

**Discussion by yonatanhakatan** [vercel/next.js#51050] [Help] [answered]
### Summary

Hi,

I have just been experimenting with next.js server components to get a feel for it and everything I have read in the official docs says that you cannot import a server component into a client component: https://nextjs.org/docs/getting-started/react-essentials#unsupported-pattern-importing-server-components-into-client-components

However, this does not align with what I am experiencing. I am seemingly able to import `Server2` inside `Client1` without next complaining as outlined below:

src/app/page.tsx:
```
import { Server1 } from "./components/Server1";

export default function Home() {
    return <Server1 />
}
```

src/app/components/Server1.tsx:
```
import { Client1 } from "./Client1";

const Server1 = () => {
  return (
    <div>
      I am server 1<br />
      <Client1 />
    </div>
  );
};

export { Server1 };
```

src/app/components/Client1.tsx:
```
"use client";
import { useState } from "react";

import { Server2 } from "./Server2";

const Client1 = () => {
  const [state, setState] = useState(false);

  return (
    <div>
      <button onClick={() => setState(true)}>I am client 1</button>
      <br />
      <Server2 />
    </div>
  );
};

export { Client1 };
```

src/app/components/Server2.tsx
```
const Server2 = async () => {
  const data = await fetch(
    "https://api.com/blah"
  );
  const data = await data.json();
  return <div>I am server 2!</div>;
};

export { Server2 };
```

I don't know if the docs are missing some finer details around this restriction or maybe i'm misunderstanding but keen to get some clarity on this.

Thanks

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.3.0: Mon Jan 30 20:42:11 PST 2023; root:xnu-8792.81.3~2/RELEASE_X86_64
    Binaries:
      Node: 18.15.0
      npm: 9.6.1
      Yarn: 1.19.1
      pnpm: N/A
    Relevant packages:
      next: 13.4.4
      eslint-config-next: 13.4.4
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
```
```


### Example

_No response_

**Comment by icyJoseph** [comment, marked as answer]
Hi,

The thing is that, all components are server components, unless, they use the `use client` directive, or, are imported into a `use client` module, in which case they are opted-in to be a Client Components, which just means that JavaScript bundle for it will be made.

> "use client" does not need to be defined in every file. The Client module boundary only needs to be defined once, at the "entry point", for all modules imported into it to be considered a Client Component. - https://nextjs.org/docs/getting-started/react-essentials#the-use-client-directive

Now the problem, and why it turns into an unsupported pattern, is, **figuring out what can be opted-in to a be a Client Component**?

Well, if it is just a synchronous function component, (classes are client components by default), that just renders some JSX, then we are good, but if the component is using Node APIs or if it is an asynchronous functions, then we'll get into trouble.

In your example, it all seems to render, until you click that `I am client 1` button. Here, try with this change (make sure to open your console):

```tsx
"use client";
import { useState } from "react";

import { Server2 } from "./Server2";

const Client1 = () => {
  const [state, setState] = useState(false);

  console.count("client 1");

  return (
    <div>
      <button onClick={() => setState(true)}>I am client 1</button>
      <p>{state ? "ok" : "not ok"}</p>
      <br />
      <Server2 />
    </div>
  );
};

export { Client1 };
```

Sorry in advance.


What you can do to prevent this accident, is called poisoning the module: https://nextjs.org/docs/getting-started/react-essentials#keeping-server-only-code-out-of-client-components-poisoning

For that you install the server-only package:

```console
npm install server-only
```

And then use it:

```ts
import 'server-only'
 
export async function getData() {
  const res = await fetch('https://external-service.com/data', {
    headers: {
      authorization: process.env.API_KEY,
    },
  })
 
  return res.json()
}

```

More here: https://nextjs.org/docs/getting-started/react-essentials#the-server-only-package


Normally, though it all crashes down anyway, but if you wanna be explicit, poisoning is a good approach.

Now with TS 5.1.x async functions as children is allowed so, one has to be extra careful!

**Comment by yonatanhakatan** [comment]
Riiiight, I see. The point `"use client" does not need to be defined in every file` was they key point I was missing. Thanks for the explanation.
